src/Features/Core/Portable/DocumentSymbols/AbstractDocumentSymbolsService.cs

```diff
@@ -0,0 +1,185 @@
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.DocumentSymbols
+{
+    internal abstract class AbstractDocumentSymbolsService : IDocumentSymbolsService
+    {
+        protected abstract ISymbol? GetSymbol(SemanticModel semanticModel, SyntaxNode node, CancellationToken cancellationToken);
+        protected abstract bool ShouldSkipSyntaxChildren(SyntaxNode node, DocumentSymbolsOptions options);
+        protected abstract bool ConsiderNestedNodesChildren(ISymbol node);
+        protected abstract DocumentSymbolInfo GetMemberInfoForType(INamedTypeSymbol type, SyntaxTree tree, ISymbolDeclarationService declarationService, CancellationToken cancellationToken);
+        protected abstract DocumentSymbolInfo CreateInfo(ISymbol symbol, SyntaxTree tree, ISymbolDeclarationService declarationService, ImmutableArray<DocumentSymbolInfo> childrenSymbols, CancellationToken cancellationToken);
+
+        public async Task<ImmutableArray<DocumentSymbolInfo>> GetSymbolsInDocumentAsync(Document document, DocumentSymbolsOptions options, CancellationToken cancellationToken)
+        {
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var tree = await document.GetRequiredSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var declarationService = document.GetRequiredLanguageService<ISymbolDeclarationService>();
+            if (options == DocumentSymbolsOptions.TypesAndMembersOnly)
+            {
+                return GetMembersInDocumentTypesAndMethodsOnly(options, semanticModel, tree, declarationService, cancellationToken);
+            }
+            else
+            {
+                return await GetMembersInDocumentFullHierarchyAsync(options, semanticModel, tree, declarationService, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        private ImmutableArray<DocumentSymbolInfo> GetMembersInDocumentTypesAndMethodsOnly(
+            DocumentSymbolsOptions options,
+            SemanticModel semanticModel,
+            SyntaxTree tree,
+            ISymbolDeclarationService declarationService,
+            CancellationToken cancellationToken)
+        {
+            using var _1 = PooledHashSet<INamedTypeSymbol>.GetInstance(out var typesInFile);
+            GetTypesInFile(semanticModel, typesInFile, cancellationToken);
+
+            using var _2 = ArrayBuilder<DocumentSymbolInfo>.GetInstance(out var typesBuilder);
+            foreach (var type in typesInFile)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                typesBuilder.Add(GetMemberInfoForType(type, tree, declarationService, cancellationToken));
+            }
+
+            typesBuilder.Sort((d1, d2) => d1.Text.CompareTo(d2.Text));
+            return typesBuilder.ToImmutable();
+
+            void GetTypesInFile(SemanticModel semanticModel, HashSet<INamedTypeSymbol> types, CancellationToken cancellationToken)
```

Comment:
can you make this static, or extract to method?  i can't really tell what is captured here and how this affects outer state.